### PR TITLE
[clang] Avoid triggering vtable instantiation for C++23 constexpr dtor

### LINKED ISSUE #102293: [clang] Clang 19 reports "arithmetic on a pointer to an incomplete type" where Clang 18 and GCC do not
https://godbolt.org/z/sYPWjWqbo

```cpp
struct _Destroy_aux {
  template <typename _ForwardIterator> static void __destroy(_ForwardIterator) {
    _ForwardIterator __first;
    ++__first;
  }
};

template <typename _ForwardIterator> void _Destroy(_ForwardIterator __last) {
  _Destroy_aux::__destroy(__last);
}

template <typename> struct allocator;

template <typename> struct allocator_traits;

template <typename _Tp> struct allocator_traits<allocator<_Tp>> {
  using pointer = _Tp *;
};

struct FileInfo;

struct __alloc_traits : allocator_traits<allocator<FileInfo>> {};

__alloc_traits::pointer _M_finish;

template <typename = int> struct vector {
  ~vector() { _Destroy(_M_finish); }
};

struct DatasetFactory {
  virtual ~DatasetFactory();
};

struct FileSystemDatasetFactory : DatasetFactory {
  vector<> v;
};
```
```
<source>:4:5: error: arithmetic on a pointer to an incomplete type 'FileInfo'
    4 |     ++__first;
      |     ^ ~~~~~~~
<source>:9:17: note: in instantiation of function template specialization '_Destroy_aux::__destroy<FileInfo *>' requested here
    9 |   _Destroy_aux::__destroy(__last);
      |                 ^
<source>:27:15: note: in instantiation of function template specialization '_Destroy<FileInfo *>' requested here
   27 |   ~vector() { _Destroy(_M_finish); }
      |               ^
<source>:34:8: note: in instantiation of member function 'vector<>::~vector' requested here
   34 | struct FileSystemDatasetFactory : DatasetFactory {
      |        ^
<source>:20:8: note: forward declaration of 'FileInfo'
   20 | struct FileInfo;
      |        ^
1 error generated.
```
Minified from Apache Arrow sources. Removing template from `vector` makes all compilers report the same error.

Is it a problem with Clang or with Arrow?

### PR DESCRIPTION
**Pull request by Fznamznon**
In C++23 anything can be constexpr, including a dtor of a class whose members and bases don't have constexpr dtors. Avoid early triggering of vtable instantiation int this case.

Fixes https://github.com/llvm/llvm-project/issues/102293

### CODE REVIEW

**Comment by Fznamznon** [comment]
I'm really in doubts that this is actually correct and useful...

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 1]
```diff
@@ -7042,12 +7042,38 @@ void Sema::CheckCompletedCXXClass(Scope *S, CXXRecordDecl *Record) {
       }
     }
 
+    bool EffectivelyConstexprDestructor = true;
+    // Avoid triggering vtable instantiation due to a dtor that is not
+    // "effectively constexpr" for better compatibility.
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaDeclCXX.cpp:7047]: I think we want more comment here (and/or a reference to the issue)

**Review by cor3ntin — COMMENTED** [review]
General direction looks good!

File: clang/test/SemaCXX/gh102293.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+struct HasVT {
+  virtual ~HasVT();
+};
+
+struct S : HasVT {
+  HasD<> v;
+};
+
```
Comment (cor3ntin) [inline clang/test/SemaCXX/gh102293.cpp:22]: What happen if you do 

`static_assert((S{}, 1));`

(and do we care that it's probably going to complain about the destructor nit being defined?)

File: clang/lib/Sema/SemaDeclCXX.cpp [hunk 3]
```diff
@@ -7042,12 +7042,38 @@ void Sema::CheckCompletedCXXClass(Scope *S, CXXRecordDecl *Record) {
...
+            return false;
+        for (const FieldDecl *FD : RD->fields())
+          if (!Check(FD->getType(), Check))
+            return false;
+        return true;
+      };
+      EffectivelyConstexprDestructor =
+          Check(QualType(Record->getTypeForDecl(), 0), Check);
+    }
+
```
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclCXX.cpp:7069]: No, unfortunately `Record->defaultedDestructorIsConstexpr()` returns false for C++23.

File: clang/test/SemaCXX/gh102293.cpp (hunk 2 above)
Comment (Fznamznon) [reply clang/test/SemaCXX/gh102293.cpp:22]: I don't think there NO tests checking that complain?

File: clang/lib/Sema/SemaDeclCXX.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (cor3ntin) [reply clang/lib/Sema/SemaDeclCXX.cpp:7070]: Yes, because it's not effectively constexpr :)

**Comment by MitalAshok** [comment]
Does this also fix #92486?

**Comment by Fznamznon** [comment]
> Does this also fix #92486?

Yes, it does.

File: clang/lib/Sema/SemaDeclCXX.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclCXX.cpp:7070]: Oh, woops, I meant it returns `true`.

[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaDeclCXX.cpp:7070]: Due to this 
https://github.com/llvm/llvm-project/blob/52126dc72c3f6f4d27e3835b0ad53e162af25e53/clang/lib/AST/DeclCXX.cpp#L1342

I think

[the remark this answers is not in the document]
Comment (cor3ntin) [reply clang/lib/Sema/SemaDeclCXX.cpp:7070]: Oh. That's unfortunate. Oh well, good enough!

**Review by cor3ntin — APPROVED** [review]
LGTM

**Comment by vient** [comment]
Will this be backported to release/19.x?

**Comment by kadircet** [comment]
hi, this seems to be triggering infinite recursion and stack exhaustion on some invalid code patterns, see https://github.com/llvm/llvm-project/issues/104802 for details.

**Pull request merged**